attribution/token_perturbation.py

```diff
@@ -0,0 +1,88 @@
+from typing import List
+
+import numpy as np
+from sklearn.neighbors import NearestNeighbors
+from transformers import PreTrainedTokenizer
+
+FIXED_REPLACEMENT_TOKEN = "the"
+
+
+def get_replacement_token(
+    token_id_to_replace: int,
+    perturbation_strategy: str,
+    embeddings: np.ndarray = None,
+    tokenizer: PreTrainedTokenizer = None,
+) -> int:
+    if perturbation_strategy == "fixed":
+        return tokenizer.encode(FIXED_REPLACEMENT_TOKEN, add_special_tokens=False)[0]
+    elif perturbation_strategy == "distant":
+        return get_increasingly_distant_token_ids(
+            token_id_to_replace, embeddings, n_tokens=4
+        )[-1]
+    elif perturbation_strategy == "nearest":
+        return get_most_similar_token_ids(
+            token_id_to_replace, embeddings, tokenizer, 2
+        )[1]
+    else:
+        raise ValueError(f"Unknown perturbation strategy: {perturbation_strategy}")
+
+
+def sort_tokens_by_similarity(
+    reference_token_id: int,
+    embeddings: np.ndarray,
+    n_tokens: int = 1,
+) -> List[int]:
+    token_embedding = embeddings[reference_token_id, :]
+
+    # Normalize the embeddings
+    normalized_embeddings = embeddings / np.linalg.norm(
+        embeddings, axis=1, keepdims=True
+    )
+    normalized_token_embedding = token_embedding / np.linalg.norm(token_embedding)
+
+    # Compute the cosine similarity
+    cosine_similarities = np.dot(normalized_embeddings, normalized_token_embedding)
+
+    # Get the indices of the top n_tokens most similar tokens
+    top_indices = np.argpartition(cosine_similarities, -n_tokens)[-n_tokens:]
+    top_indices = top_indices[np.argsort(cosine_similarities[top_indices])][::-1]
+
+    return top_indices
+
+
+def get_most_similar_token_ids(
+    token_id: int,
+    embeddings: np.ndarray,
+    tokenizer: PreTrainedTokenizer,
+    n_tokens: int = 1,
+) -> List[int]:
+    token_embedding = embeddings[token_id, :]
+
+    # Fit the NearestNeighbors model to the embeddings
+    nbrs = NearestNeighbors(n_neighbors=n_tokens, algorithm="ball_tree").fit(embeddings)
+
+    # Find the nearest neighbor
+    indices = nbrs.kneighbors(token_embedding.reshape(1, -1), return_distance=False)
+
+    return indices.flatten().tolist()
+
+
+def get_increasingly_distant_token_ids(
+    token_id: int,
+    embeddings: np.ndarray,
+    n_tokens: int = 1,
+) -> List[int]:
+    if n_tokens > 4:
+        raise ValueError("n_tokens cannot be more than 4")
+
+    # Define the sequence of positions
+    positions = [10**i for i in range(1, n_tokens + 1)]
```

Comment:
what’s the rationale behind get tokens at positions [10, 100, 1000, 10000]? Seems to work, just curious

Reply:
The idea is to sort all the vocabulary by similarity, and then get exponentially more distant tokens. 

Initially, we thought about doing multiple perturbations per token. In this case, it makes sense that the multiple perturbations are exponentially more distant, as you don't want to change a token with the 2nd, 3rd, 4th, and 5th more distant tokens as those would be quite similar to the original one.

We are now just using this method to get a `distant` token - the 10.000th most distant token:
```
def get_replacement_token(
    token_id_to_replace: int,
    perturbation_strategy: str,
    embeddings: np.ndarray = None,
    tokenizer: PreTrainedTokenizer = None,
) -> int:
    if perturbation_strategy == "fixed":
        return tokenizer.encode(FIXED_REPLACEMENT_TOKEN, add_special_tokens=False)[0]
    elif perturbation_strategy == "distant":
        return get_increasingly_distant_token_ids(
            token_id_to_replace, embeddings, n_tokens=4
        )[-1]
...
```        
because of `n_tokens=4` and `[-1]`

Reply:
The issue with doing _multiple_ perturbations per token is that you introduce a second dimension across which "attribution" can be computed: do you compute an attribution score by looking at how different two outputs are, by looking at "how different" the replaced token is, or an average of the two?

Reply:
Sounds good! Yeah, it's easy to get the nearest, furthest and k-th furthest token with the provided functions. 
One thing to note though, is that the most similar tokens are usually variations of the same token, with different capitalization:

<img width="951" alt="image" src="https://github.com/leap-laboratories/llm-attribution/assets/34447985/6b4f5371-18f6-42a5-8bf9-100c2b8fff2b">